Allow functors as reducers for nested team parallel reduce

ldh4 opened the pull request:
Resolves #6317.

This PR allows functors to be used as reducers in nested parallel reduction, specifically for `parallel_reduce` that takes in `TeamThreadRange`, `ThreadVectorRange` or `TeamVectorRange`.
This capability is already available for reductions with `Range/MDRange/Team` policies as noted in #6774.

Changes in this PR convert nested parallel reduces to use `FunctorAnalysis::Reducer` in order to avoid explicitly checking for the availability of `init` or `final` implementation in custom reducers.

Also, most of nested `parallel_reduce` weren't calling `final` even when it is available in a custom reducer. So that was addressed in this PR as well.

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -265,39 +267,42 @@ class CudaTeamMember {
+        Impl::in_place_shfl(tmp2, tmp, 0, blockDim.x, mask);
+        value = tmp2;
+
+        if constexpr (is_reducer_v<ReducerType>) {
+          reducer.reference() = tmp2;
+        }))
```
Why aren't we using `final` here if defined?

ldh4 replied:
Fixed to call `final` through the `FunctorAnalysis::Reducer` interface.
It seems that `final` wasn't being called for any of nested team `parallel_reduce`.

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -667,15 +723,41 @@ parallel_reduce(Impl::ThreadVectorRangeBoundariesStruct<
+          functor_analysis_type::has_join_member_function &&
+          functor_analysis_type::has_init_member_function;
+
+      using ReducerSelector =
+          typename Kokkos::Impl::if_c<is_reducer_closure, Closure,
+                                      Sum<ValueType>>::type;
```
See above.

ldh4 replied:
Removed and replaced by `FunctorAnalysis::Reducer` usage.

masterleinad commented on core/unit_test/TestTeam.hpp:
```diff
@@ -1753,4 +1753,133 @@ struct TestRepeatedTeamReduce {
+    dst.value *= src.value;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void final(value_type &) const {}
```
We should test with a non-trivial `final` as well.

ldh4 replied:
Modified the test to have something testable in `final`.

masterleinad reviewed (commented):
I only commented on the `Cuda` backend but the same comments apply to all other backends.

ldh4 commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -518,16 +523,41 @@ parallel_reduce(const Impl::TeamThreadRangeBoundariesStruct<
+
+      constexpr bool is_reducer_closure =
+          functor_analysis_type::has_join_member_function &&
+          functor_analysis_type::has_init_member_function;
```
Removed and replaced by `FunctorAnalysis::Reducer` usage.

ldh4 commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -518,16 +523,41 @@ parallel_reduce(const Impl::TeamThreadRangeBoundariesStruct<
+
+      using ReducerSelector =
+          typename Kokkos::Impl::if_c<is_reducer_closure, Closure,
+                                      Sum<ValueType>>::type;
```
Converted to use `FunctorAnalysis::Reducer`.

ldh4 commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -573,18 +603,44 @@ parallel_reduce(const Impl::TeamVectorRangeBoundariesStruct<
+
+      using ReducerSelector =
+          typename Kokkos::Impl::if_c<is_reducer_closure, Closure,
+                                      Sum<ValueType>>::type;
```
Removed and replaced by `FunctorAnalysis::Reducer` usage.

ldh4 commented:
Retest this please.

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -186,22 +186,23 @@ class CudaTeamMember {
   KOKKOS_INLINE_FUNCTION std::enable_if_t<is_reducer<ReducerType>::value>
-  team_reduce(ReducerType const& reducer,
-              typename ReducerType::value_type& value) const noexcept {
+  team_reduce(ReducerType const& reducer, ValueType& value) const noexcept {
```
What if `typename ReducerType::value_type` and `ValueType` don't match?

ldh4 replied:
Reverted to use `ReducerType::value_type`.

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -186,22 +186,23 @@ class CudaTeamMember {
-             wrapped_reducer(reducer);
-         cuda_intra_block_reduction(value, wrapped_reducer, blockDim.y);
-         reducer.reference() = value;))
+        (cuda_intra_block_reduction(value, reducer, blockDim.y);))
```
Why is the reducer not wrapped anymore? This function is still user-facing, right?

Why don't you need to have
```
reducer.reference() = value;
```
anymore? Don't you need to assign in the other overload then?

ldh4 replied:
The assumption I was operating under was that the team member class and its member functions are not necessarily user-facing, because they are placed in `impl` namespace.
But, seeing that they are defined to be part of the public API in the [documentation](https://kokkos.org/kokkos-core-wiki/API/core/policies/TeamHandleConcept.html), it does make some of the changes I made in this PR invasive. I'll revert and fix those.

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -262,13 +263,18 @@ class CudaTeamMember {
   KOKKOS_INLINE_FUNCTION static std::enable_if_t<is_reducer<ReducerType>::value>
-  vector_reduce(ReducerType const& reducer,
-                typename ReducerType::value_type& value) {
+  vector_reduce(ReducerType const& reducer, ValueType& value) {
```
Same here. SInce this is user-facing we shouldn't expect a wrapped reducer to be passed in but wrap the reducer in this function instead.

ldh4 replied:
Changed so that existing function signatures remain unchanged. Instead introduced a function that will take in a wrapped reducer.

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -487,14 +493,23 @@ parallel_reduce(const Impl::TeamThreadRangeBoundariesStruct<
-       loop_boundaries.member.team_reduce(reducer, value);))
+       loop_boundaries.member.team_reduce(wrapped_reducer, value);
+       wrapped_reducer.final(&value); reducer.reference() = value;
+
```
```suggestion
```

masterleinad reviewed (commented):
Just commenting on the changes to the `Cuda` backend.

masterleinad commented:
Looks like the `HPX` backend doesn't work with the changes here.

ldh4 commented:
Rebased and squashed to avoid having too many stale commits.
Reverted few changes on user-facing functions so that their function signatures remain as described in kokkos docs.

masterleinad commented:
This still failing for a bunch of backends.

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -487,14 +519,21 @@ parallel_reduce(const Impl::TeamThreadRangeBoundariesStruct<
+       using wrapped_reducer_type = typename functor_analysis_type::Reducer;
 
-       reducer.init(value);
+       wrapped_reducer_type wrapped_reducer(reducer); value_type value{};
```
We don't really need the explicit initialization here. If we omit it, there is a higher chance of catching bugs in the implementation.
```suggestion
       wrapped_reducer_type wrapped_reducer(reducer); value_type value;
```

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -548,16 +596,27 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<Kokkos::is_reducer<ReducerType>::value>
-                                      threadIdx.y * blockDim.x + threadIdx.x;
-                            i < loop_boundaries.end;
-                            i += blockDim.y * blockDim.x) { closure(i, value); }
+       wrapped_reducer_type wrapped_reducer(reducer); value_type value{};
```
```suggestion
       wrapped_reducer_type wrapped_reducer(reducer); value_type value;
```

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -573,18 +632,27 @@ parallel_reduce(const Impl::TeamVectorRangeBoundariesStruct<
-                                      threadIdx.y * blockDim.x + threadIdx.x;
-                            i < loop_boundaries.end;
-                            i += blockDim.y * blockDim.x) { closure(i, val); }
+       wrapped_reducer_type wrapped_reducer(closure); value_type value{};
```
```suggestion
       wrapped_reducer_type wrapped_reducer(closure); value_type value;
```

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -632,13 +700,22 @@ parallel_reduce(Impl::ThreadVectorRangeBoundariesStruct<
+          ReducerType, value_type>;
+      using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+      wrapped_reducer_type wrapped_reducer(reducer); value_type value{};
```
```suggestion
      wrapped_reducer_type wrapped_reducer(reducer); value_type value;
```

masterleinad commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -667,15 +744,26 @@ parallel_reduce(Impl::ThreadVectorRangeBoundariesStruct<
+      using value_type           = typename wrapped_reducer_type::value_type;
 
-       Impl::CudaTeamMember::vector_reduce(Kokkos::Sum<ValueType>(result));
+      wrapped_reducer_type wrapped_reducer(closure); value_type value{};
```
```suggestion
      wrapped_reducer_type wrapped_reducer(closure); value_type value;
```

masterleinad commented on core/src/HIP/Kokkos_HIP_Team.hpp:
```diff
@@ -479,15 +511,26 @@ parallel_reduce(const Impl::TeamThreadRangeBoundariesStruct<
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/HIP/Kokkos_HIP_Team.hpp:
```diff
@@ -508,24 +551,24 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!Kokkos::is_reducer<ValueType>::value>
+       using value_type           = typename wrapped_reducer_type::value_type;
 
-  reducer.init(reducer.reference());
+       wrapped_reducer_type wrapped_reducer(closure); value_type value{};
```
```suggestion
       wrapped_reducer_type wrapped_reducer(closure); value_type value;
```

masterleinad commented on core/src/HIP/Kokkos_HIP_Team.hpp:
```diff
@@ -620,16 +663,26 @@ parallel_reduce(const Impl::TeamVectorRangeBoundariesStruct<
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/HIP/Kokkos_HIP_Team.hpp:
```diff
@@ -642,25 +695,27 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!Kokkos::is_reducer<ValueType>::value>
+       using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+       using value_type           = typename wrapped_reducer_type::value_type;
+
+       wrapped_reducer_type wrapped_reducer(closure); value_type value{};
```
```suggestion
       wrapped_reducer_type wrapped_reducer(closure); value_type value;
```

masterleinad commented on core/src/HIP/Kokkos_HIP_Team.hpp:
```diff
@@ -706,14 +761,26 @@ parallel_reduce(Impl::ThreadVectorRangeBoundariesStruct<
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/HIP/Kokkos_HIP_Team.hpp:
```diff
@@ -737,20 +804,24 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!is_reducer<ValueType>::value>
-       i += blockDim.x) {
-    closure(i, result);
-  }
+       wrapped_reducer_type wrapped_reducer(closure); value_type value{};
```
```suggestion
       wrapped_reducer_type wrapped_reducer(closure); value_type value;
```

masterleinad commented on core/src/HPX/Kokkos_HPX.hpp:
```diff
@@ -1786,11 +1786,24 @@ KOKKOS_INLINE_FUNCTION void parallel_reduce(
+  using value_type           = typename wrapped_reducer_type::value_type;
+
+  wrapped_reducer_type wrapped_reducer(lambda);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/HPX/Kokkos_HPX.hpp:
```diff
@@ -1824,14 +1837,26 @@ KOKKOS_INLINE_FUNCTION void parallel_reduce(
+  using value_type           = typename wrapped_reducer_type::value_type;
+
+  wrapped_reducer_type wrapped_reducer(lambda);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/HPX/Kokkos_HPX.hpp:
```diff
@@ -1840,11 +1865,24 @@ KOKKOS_INLINE_FUNCTION void parallel_reduce(
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/HPX/Kokkos_HPX.hpp:
```diff
@@ -1853,14 +1891,27 @@ KOKKOS_INLINE_FUNCTION void parallel_reduce(
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/SYCL/Kokkos_SYCL_Team.hpp:
```diff
@@ -531,8 +555,16 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<Kokkos::is_reducer<ReducerType>::value>
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/SYCL/Kokkos_SYCL_Team.hpp:
```diff
@@ -557,20 +591,28 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!Kokkos::is_reducer<ValueType>::value>
 
-  reducer.init(reducer.reference());
+  wrapped_reducer_type wrapped_reducer(closure);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/SYCL/Kokkos_SYCL_Team.hpp:
```diff
@@ -657,8 +699,16 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<Kokkos::is_reducer<ReducerType>::value>
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/SYCL/Kokkos_SYCL_Team.hpp:
```diff
@@ -670,19 +720,28 @@ parallel_reduce(const Impl::TeamVectorRangeBoundariesStruct<
 
-  reducer.init(reducer.reference());
+  wrapped_reducer_type wrapped_reducer(closure);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/SYCL/Kokkos_SYCL_Team.hpp:
```diff
@@ -746,16 +807,27 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<is_reducer<ReducerType>::value>
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/SYCL/Kokkos_SYCL_Team.hpp:
```diff
@@ -774,16 +846,27 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!is_reducer<ValueType>::value>
+  using value_type           = typename wrapped_reducer_type::value_type;
+
+  wrapped_reducer_type wrapped_reducer(closure);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/Threads/Kokkos_Threads_Team.hpp:
```diff
@@ -887,35 +900,51 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!Kokkos::is_reducer<ValueType>::value>
+  using value_type           = typename wrapped_reducer_type::value_type;
+
+  wrapped_reducer_type wrapped_reducer(lambda);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/Threads/Kokkos_Threads_Team.hpp:
```diff
@@ -887,35 +900,51 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!Kokkos::is_reducer<ValueType>::value>
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/Threads/Kokkos_Threads_Team.hpp:
```diff
@@ -950,23 +979,49 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!Kokkos::is_reducer<ValueType>::value>
+  using value_type           = typename wrapped_reducer_type::value_type;
+
+  wrapped_reducer_type wrapped_reducer(lambda);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/Threads/Kokkos_Threads_Team.hpp:
```diff
@@ -950,23 +979,49 @@ KOKKOS_INLINE_FUNCTION std::enable_if_t<!Kokkos::is_reducer<ValueType>::value>
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/impl/Kokkos_HostThreadTeam.hpp:
```diff
@@ -781,15 +797,25 @@ KOKKOS_INLINE_FUNCTION
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/impl/Kokkos_HostThreadTeam.hpp:
```diff
@@ -799,17 +825,24 @@ KOKKOS_INLINE_FUNCTION
+  using value_type           = typename wrapped_reducer_type::value_type;
+
+  wrapped_reducer_type wrapped_reducer(closure);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/impl/Kokkos_HostThreadTeam.hpp:
```diff
@@ -853,11 +886,23 @@ KOKKOS_INLINE_FUNCTION
+  using value_type           = typename wrapped_reducer_type::value_type;
+
+  wrapped_reducer_type wrapped_reducer(lambda);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad commented on core/src/impl/Kokkos_HostThreadTeam.hpp:
```diff
@@ -867,11 +912,23 @@ KOKKOS_INLINE_FUNCTION
+  using wrapped_reducer_type = typename functor_analysis_type::Reducer;
+
+  wrapped_reducer_type wrapped_reducer(reducer);
+  value_type value{};
```
```suggestion
  value_type value;
```

masterleinad reviewed (commented):
Looks pretty good but I would avoid the explicit initialization.

diehlpk commented:
> Looks like the `HPX` backend doesn't work with the changes here.

we have two failing tests for the HPX backend already. That might not be related to this pull request.

diehlpk commented:
@masterleinad is there some specific backend, I should look at?

ldh4 commented:
> > Looks like the `HPX` backend doesn't work with the changes here.
> 
> we have two failing tests for the HPX backend already. That might not be related to this pull request.

I think there was indeed an error with changes in this PR at the time that caused the CI to fail for the HPX backend.
I believe those errors were fixed with more recent changes in this PR.

diehlpk reviewed (approved):
LGTM!
